File: bots/bot.js
/**
 * Poker Engine — Node.js Bot Boilerplate
 * =======================================
 * Copy this file, implement the `decide()` function, and deploy.
 *
 * Run: node bot.js [port]
 *
 * Your only job: implement decide(state) at the bottom of this file.
 * Everything else is handled for you.
 */

'use strict';
const http = require('http');
const PORT = process.argv[2] || 3001;

// ─────────────────────────────────────────────────────────────
// ENTITIES  —  typed wrappers around the raw JSON payload
// ─────────────────────────────────────────────────────────────

class Card {
  constructor(str) {
    this.raw = str;
    this.hidden = str === '??';
    if (!this.hidden) {
      this.suit = str.slice(-1);           // ♠ ♥ ♦ ♣
      this.rank = str.slice(0, -1);        // A K Q J 10 9 ... 2
      this.value = Card.rankValue(this.rank);
      this.isRed = this.suit === '♥' || this.suit === '♦';
    }
  }
  static rankValue(rank) {
    const map = { '2':2,'3':3,'4':4,'5':5,'6':6,'7':7,'8':8,'9':9,'10':10,'J':11,'Q':12,'K':13,'A':14 };
    return map[rank] || 0;
  }
  toString() { return this.raw; }
}

class BestHand {
  constructor(data) {
    this.name = data.name;           // e.g. "FLUSH"
    this.cards = data.cards.map(c => new Card(c));
    this.rank = BestHand.RANKS[data.name] || 0;
  }
  static RANKS = {
    HIGH_CARD:0, ONE_PAIR:1, TWO_PAIR:2, THREE_OF_A_KIND:3,
    STRAIGHT:4, FLUSH:5, FULL_HOUSE:6, FOUR_OF_A_KIND:7,
    STRAIGHT_FLUSH:8, ROYAL_FLUSH:9,
  };
  isAt(name) { return this.name === name; }
  isAtLeast(name) { return this.rank >= BestHand.RANKS[name]; }
}

class ActionOptions {
  constructor(data) {
    this.canCheck = data.canCheck;
    this.toCall   = data.toCall;
    this.minRaise = data.minRaise;
    this.maxRaise = data.maxRaise;
  }
  /** Pot odds as a fraction (0–1). Call if your equity exceeds this. */
  potOdds(pot) { return this.toCall / (pot + this.toCall) || 0; }
  canRaise()   { return this.maxRaise > 0; }
  canCall()    { return this.toCall > 0; }
}

class Player {
  constructor(data) {
    this.name         = data.name;
    this.chips        = data.chips;
    this.bet          = data.bet;
    this.folded       = data.folded;
    this.allIn        = data.allIn;
    this.position     = data.position;
    this.disconnected = data.disconnected;
    this.isActive     = !data.folded && !data.allIn && !data.disconnected && data.chips > 0;
  }
}

class YouState {
  constructor(data) {
    this.name      = data.name;
    this.chips     = data.chips;
    this.holeCards = data.holeCards.map(c => new Card(c));
    this.bet       = data.bet;
    this.position  = data.position;
    this.bestHand  = data.bestHand ? new BestHand(data.bestHand) : null;
  }

  /** True if playing from late position (CO, BTN, BTN/SB) */
  inPosition()  { return ['BTN','CO','BTN/SB'].includes(this.position); }
  isBlind()     { return ['SB','BB','BTN/SB'].includes(this.position); }
  isDealer()    { return ['BTN','BTN/SB'].includes(this.position); }

  /** Effective stack as multiple of big blind */
  stackInBBs(bigBlind) { return this.chips / bigBlind; }
}

class TableState {
  constructor(data) {
    this.pot            = data.pot;
    this.currentBet     = data.currentBet;
    this.communityCards = data.communityCards.map(c => new Card(c));
    this.smallBlind     = data.smallBlind;
    this.bigBlind       = data.bigBlind;
    this.ante           = data.ante;
  }
  /** Total pot if we call (for pot odds calculation) */
  potAfterCall(toCall) { return this.pot + toCall; }
  hasFlop()  { return this.communityCards.length >= 3; }
  hasTurn()  { return this.communityCards.length >= 4; }
  hasRiver() { return this.communityCards.length === 5; }
}

class GameState {
  constructor(payload) {
    this.gameId     = payload.gameId;
    this.handNumber = payload.handNumber;
    this.stage      = payload.stage;   // 'pre-flop' | 'flop' | 'turn' | 'river'
    this.you        = new YouState(payload.you);
    this.action     = new ActionOptions(payload.action);
    this.table      = new TableState(payload.table);
    this.players    = payload.players.map(p => new Player(p));
  }

  /** Players still in the hand (not folded, not disconnected) */
  activePlayers() { return this.players.filter(p => !p.folded && !p.disconnected); }

  /** Number of opponents still in the hand */
  opponentCount() { return this.activePlayers().filter(p => p.name !== this.you.name).length; }

  isPreFlop() { return this.stage === 'pre-flop'; }
  isFlop()    { return this.stage === 'flop'; }
  isTurn()    { return this.stage === 'turn'; }
  isRiver()   { return this.stage === 'river'; }
}

// ─────────────────────────────────────────────────────────────
// ACTIONS  —  helpers for building valid responses
// ─────────────────────────────────────────────────────────────

const Action = {
  fold:  ()      => ({ type: 'fold' }),
  check: ()      => ({ type: 'check' }),
  call:  ()      => ({ type: 'call' }),

  /**
   * Raise by `amount` additional chips (on top of the call).
   * Amount is automatically clamped to [minRaise, maxRaise].
   */
  raiseBy: (amount, actionOptions) => {
    const clamped = actionOptions
      ? Math.min(Math.max(amount, actionOptions.minRaise), actionOptions.maxRaise)
      : amount;
    return { type: 'raise', amount: Math.floor(clamped) };
  },

  /** Go all-in */
  allIn: (actionOptions) => ({
    type: 'raise',
    amount: actionOptions ? actionOptions.maxRaise : 999999,
  }),

  /** Pot-sized raise (raise by the current pot size) */
  potSized: (pot, actionOptions) => Action.raiseBy(pot, actionOptions),

  /** Half-pot raise */
  halfPot: (pot, actionOptions) => Action.raiseBy(Math.floor(pot / 2), actionOptions),

  /** Check if possible, otherwise call */
  checkOrCall: (actionOptions) =>
    actionOptions.canCheck ? Action.check() : Action.call(),
};

// ─────────────────────────────────────────────────────────────
// HELPERS
// ─────────────────────────────────────────────────────────────

/**
 * Pot odds — fraction of the total pot you need to invest to call.
 * If your hand equity exceeds this, calling is profitable long-term.
 * Example: toCall=100, pot=300 → 0.25 → need >25% equity to call.
 */
function potOdds(toCall, pot) {
  if (toCall === 0) return 0;
  return toCall / (pot + toCall);
}

/**
 * Pre-flop hand strength by position.
 *
 * Returns 0–1. Higher = stronger / more playable.
 * Takes position into account: late position widens the playable range.
 *
 * Approach: score each hole card individually then combine.
 * Pairs, suited connectors, and broadway cards score higher.
 */
function preFlopStrength(holeCards, position) {
  if (holeCards.length < 2 || holeCards[0].hidden) return 0.1;

  const [a, b] = holeCards.sort((x, y) => y.value - x.value); // high card first
  const isPair     = a.rank === b.rank;
  const isSuited   = a.suit === b.suit;
  const gap        = a.value - b.value;        // 0 = connected, 1 = one-gapper
  const isBroadway = a.value >= 10 && b.value >= 10;
  const hasAce     = a.value === 14;
  const inLate     = ['BTN', 'CO', 'BTN/SB'].includes(position);
  const inMid      = ['HJ', 'MP', 'MP+1'].includes(position);

  let score = 0;

  // Pairs
  if (isPair) {
    if (a.value >= 10) score = 0.95;       // JJ+
    else if (a.value >= 7) score = 0.70;   // 77–TT
    else score = 0.50;                     // 22–66 (set mining)
  }
  // Broadway (AK, AQ, KQ, etc.)
  else if (isBroadway) {
    if (a.value === 14 && b.value === 13) score = isSuited ? 0.90 : 0.85; // AK
    else if (a.value === 14) score = isSuited ? 0.80 : 0.72;              // AQ, AJ, AT
    else score = isSuited ? 0.68 : 0.58;                                   // KQ, KJ, QJ etc.
  }
  // Ace with non-broadway kicker
  else if (hasAce) {
    score = isSuited ? 0.60 : 0.45;       // A9s, A5s, A2o etc.
  }
  // Suited connectors and one-gappers
  else if (isSuited && gap <= 2) {
    score = a.value >= 7 ? 0.55 : 0.42;   // High suited connectors > low
  }
  // Offsuit connectors
  else if (gap === 0) {
    score = a.value >= 8 ? 0.42 : 0.30;
  }
  else {
    score = 0.15; // trash
  }

  // Position bonus: late position can profitably play wider ranges
  if (inLate) score = Math.min(score + 0.12, 1.0);
  else if (inMid) score = Math.min(score + 0.06, 1.0);

  return score;
}

/**
 * Post-flop hand strength from bestHand.
 * Adjusted for number of opponents — a pair is weaker against 4 opponents.
 */
function postFlopStrength(bestHand, opponentCount) {
  if (!bestHand) return 0.1;
  const base = {
    HIGH_CARD:0.08, ONE_PAIR:0.30, TWO_PAIR:0.52, THREE_OF_A_KIND:0.65,
    STRAIGHT:0.76, FLUSH:0.82, FULL_HOUSE:0.90, FOUR_OF_A_KIND:0.97,
    STRAIGHT_FLUSH:0.99, ROYAL_FLUSH:1.0,
  }[bestHand.name] || 0.1;

  // Discount for more opponents — more people = more likely someone has you beat
  const discount = Math.max(0, (opponentCount - 1) * 0.05);
  return Math.max(base - discount, 0.02);
}

/**
 * Stack-to-pot ratio (SPR).
 * Low SPR (<4): commit with top pair or better.
 * High SPR (>10): need strong hands to build pot.
 */
function spr(chips, pot) {
  return pot > 0 ? chips / pot : 99;
}

/**
 * Effective stack in big blinds. Under 10BB → push/fold territory.
 */
function stackBBs(chips, bigBlind) {
  return bigBlind > 0 ? chips / bigBlind : 99;
}

/**
 * Bet sizing — returns chips to raise by (additional on top of call).
 * Scales with hand strength: stronger hands bet more.
 */
function betSize(strength, pot, action) {
  let fraction;
  if (strength >= 0.90) fraction = 1.0;       // pot-sized with monsters
  else if (strength >= 0.70) fraction = 0.75; // 3/4 pot with strong hands
  else if (strength >= 0.55) fraction = 0.50; // half pot with medium hands
  else fraction = 0.33;                        // small probe bet
  return Action.raiseBy(Math.floor(pot * fraction), action);
}

// ─────────────────────────────────────────────────────────────
// YOUR STRATEGY  —  implement this function
// ─────────────────────────────────────────────────────────────

/**
 * decide(state) — your bot's brain.
 *
 * @param {GameState} state  — the full game state
 * @returns {object}          — an Action: fold / check / call / raise
 *
 * Replace or extend this function with your own strategy.
 * The default implementation uses position-aware pre-flop ranges
 * and pot odds post-flop. It's a reasonable starting point but
 * definitely beatable — that's the point.
 */
function decide(state) {
  const { you, action, table } = state;
  const opponents = state.opponentCount();
  const bbSize    = table.bigBlind;
  const myStack   = you.chips;
  const myBBs     = stackBBs(myStack, bbSize);

  // ── SHORT STACK: push/fold under 10 big blinds ──────────────
  if (myBBs < 10 && !action.canCheck) {
    const pfStr = preFlopStrength(you.holeCards, you.position);
    // Push any decent hand, fold the rest — standard short-stack ICM play
    if (pfStr >= 0.50) return Action.allIn(action);
    return Action.fold();
  }

  // ── PRE-FLOP ─────────────────────────────────────────────────
  if (state.isPreFlop()) {
    const pfStr = preFlopStrength(you.holeCards, you.position);
    const toCall = action.toCall;
    const raiseSize = bbSize * 2.5; // standard open raise

    // No action to us yet (we can check, or we're in BB and nobody raised)
    if (action.canCheck) {
      if (pfStr >= 0.55) return Action.raiseBy(raiseSize, action); // open raise
      return Action.check();
    }

    // Facing a raise
    const callFraction = toCall / myStack;
    if (pfStr >= 0.85) {
      // Premium hand — 3-bet
      return Action.raiseBy(toCall * 3, action);
    }
    if (pfStr >= 0.65 && callFraction < 0.10) {
      // Good hand, call if not too expensive
      return Action.call();
    }
    if (pfStr >= 0.50 && callFraction < 0.05 && you.inPosition()) {
      // Speculative hand in position with good price
      return Action.call();
    }
    return Action.fold();
  }

  // ── POST-FLOP ─────────────────────────────────────────────────
  const strength = postFlopStrength(you.bestHand, opponents);
  const odds     = potOdds(action.toCall, table.pot);
  const currentSPR = spr(myStack, table.pot);

  // Facing a bet or raise
  if (!action.canCheck) {
    // With very strong hands and low SPR, consider going all-in
    if (strength >= 0.82 && currentSPR < 4 && action.canRaise()) {
      return Action.allIn(action);
    }
    // Raise with strong hands if there's value
    if (strength >= 0.72 && action.canRaise() && action.toCall < myStack * 0.25) {
      return betSize(strength, table.pot, action);
    }
    // Call if equity justifies it (with a small buffer)
    if (strength >= odds + 0.08) {
      return Action.call();
    }
    return Action.fold();
  }

  // Nobody bet — we can check
  if (strength >= 0.55 && action.canRaise()) {
    // Bet for value (or as a probe with medium hands in position)
    if (strength >= 0.65 || you.inPosition()) {
      return betSize(strength, table.pot, action);
    }
  }
  // Check with weak hands, out of position, or on dry boards
  return Action.check();
}

// ─────────────────────────────────────────────────────────────
// SERVER  —  do not modify below this line
// ─────────────────────────────────────────────────────────────

const server = http.createServer(async (req, res) => {
  // CORS
  res.setHeader('Access-Control-Allow-Origin', '*');

  if (req.method === 'GET' && req.url === '/health') {
    res.writeHead(200, { 'Content-Type': 'application/json' });
    return res.end(JSON.stringify({ status: 'ok', bot: 'Node.js Boilerplate' }));
  }

  if (req.method === 'POST' && req.url === '/action') {
    let body = '';
    req.on('data', chunk => body += chunk);
    req.on('end', async () => {
      try {
        const payload = JSON.parse(body);
        const state = new GameState(payload);

        console.log(
          `[Hand ${state.handNumber}] ${state.stage.toUpperCase()}` +
          ` | ${state.you.holeCards.join(' ')}` +
          ` | Pot: ${state.table.pot}` +
          ` | To call: ${state.action.toCall}` +
          (state.you.bestHand ? ` | Best: ${state.you.bestHand.name}` : '')
        );

        const result = await Promise.resolve(decide(state));

        console.log(`  → ${result.type}${result.amount !== undefined ? ' ' + result.amount : ''}`);

        res.writeHead(200, { 'Content-Type': 'application/json' });
        res.end(JSON.stringify(result));
      } catch (err) {
        console.error('[Bot error]', err.message);
        // Always return a valid action even on error — fold is safe
        res.writeHead(200, { 'Content-Type': 'application/json' });
        res.end(JSON.stringify({ type: 'fold' }));
      }
    });
    return;
  }

  res.writeHead(404);
  res.end('Not found');
});

server.listen(PORT, () => {
  console.log(`🤖 Bot running on http://localhost:${PORT}`);
  console.log(`   POST /action  — receives game state, returns action`);
  console.log(`   GET  /health  — health check`);
});

File: bots/bot.py
"""
Poker Engine — Python Bot Boilerplate
======================================
Copy this file, implement the `decide()` function, and deploy.

Run: python bot.py [port]
Requires: Python 3.8+, no external dependencies.

Your only job: implement decide(state) at the bottom of this file.
"""

import json
import sys
from dataclasses import dataclass, field
from enum import Enum
from http.server import BaseHTTPRequestHandler, HTTPServer
from typing import List, Optional

PORT = int(sys.argv[1]) if len(sys.argv) > 1 else 3001


# ─────────────────────────────────────────────────────────────
# ENTITIES
# ─────────────────────────────────────────────────────────────

RANK_VALUES = {
    '2': 2, '3': 3, '4': 4, '5': 5, '6': 6, '7': 7,
    '8': 8, '9': 9, '10': 10, 'J': 11, 'Q': 12, 'K': 13, 'A': 14,
}

HAND_RANKS = {
    'HIGH_CARD': 0, 'ONE_PAIR': 1, 'TWO_PAIR': 2,
    'THREE_OF_A_KIND': 3, 'STRAIGHT': 4, 'FLUSH': 5,
    'FULL_HOUSE': 6, 'FOUR_OF_A_KIND': 7,
    'STRAIGHT_FLUSH': 8, 'ROYAL_FLUSH': 9,
}


@dataclass
class Card:
    raw: str
    hidden: bool = False
    suit: str = ''
    rank: str = ''
    value: int = 0
    is_red: bool = False

    @classmethod
    def from_str(cls, s: str) -> 'Card':
        if s == '??':
            return cls(raw=s, hidden=True)
        suit = s[-1]
        rank = s[:-1]
        return cls(
            raw=s,
            suit=suit,
            rank=rank,
            value=RANK_VALUES.get(rank, 0),
            is_red=suit in ('♥', '♦'),
        )

    def __str__(self) -> str:
        return self.raw


@dataclass
class BestHand:
    name: str
    cards: List[Card]
    rank: int = 0

    @classmethod
    def from_dict(cls, d: dict) -> 'BestHand':
        obj = cls(
            name=d['name'],
            cards=[Card.from_str(c) for c in d['cards']],
        )
        obj.rank = HAND_RANKS.get(d['name'], 0)
        return obj

    def is_at_least(self, name: str) -> bool:
        return self.rank >= HAND_RANKS.get(name, 0)


@dataclass
class ActionOptions:
    can_check: bool
    to_call: int
    min_raise: int
    max_raise: int

    def pot_odds(self, pot: int) -> float:
        if self.to_call == 0:
            return 0.0
        return self.to_call / (pot + self.to_call)

    def can_raise(self) -> bool:
        return self.max_raise > 0

    def can_call(self) -> bool:
        return self.to_call > 0


@dataclass
class Player:
    name: str
    chips: int
    bet: int
    folded: bool
    all_in: bool
    position: str
    disconnected: bool

    @property
    def is_active(self) -> bool:
        return not self.folded and not self.all_in and not self.disconnected and self.chips > 0


@dataclass
class YouState:
    name: str
    chips: int
    hole_cards: List[Card]
    bet: int
    position: str
    best_hand: Optional[BestHand]

    LATE_POSITIONS = {'BTN', 'CO', 'BTN/SB'}
    BLIND_POSITIONS = {'SB', 'BB', 'BTN/SB'}

    def in_position(self) -> bool:
        return self.position in self.LATE_POSITIONS

    def is_blind(self) -> bool:
        return self.position in self.BLIND_POSITIONS

    def stack_in_bbs(self, big_blind: int) -> float:
        return self.chips / big_blind if big_blind else 0


@dataclass
class TableState:
    pot: int
    current_bet: int
    community_cards: List[Card]
    small_blind: int
    big_blind: int
    ante: int

    def has_flop(self) -> bool:
        return len(self.community_cards) >= 3

    def has_turn(self) -> bool:
        return len(self.community_cards) >= 4

    def has_river(self) -> bool:
        return len(self.community_cards) == 5


@dataclass
class GameState:
    game_id: str
    hand_number: int
    stage: str
    you: YouState
    action: ActionOptions
    table: TableState
    players: List[Player]

    def active_players(self) -> List[Player]:
        return [p for p in self.players if not p.folded and not p.disconnected]

    def opponent_count(self) -> int:
        return len([p for p in self.active_players() if p.name != self.you.name])

    def is_pre_flop(self) -> bool: return self.stage == 'pre-flop'
    def is_flop(self) -> bool:     return self.stage == 'flop'
    def is_turn(self) -> bool:     return self.stage == 'turn'
    def is_river(self) -> bool:    return self.stage == 'river'

    @classmethod
    def from_dict(cls, d: dict) -> 'GameState':
        you_d = d['you']
        tbl_d = d['table']
        act_d = d['action']
        return cls(
            game_id=d['gameId'],
            hand_number=d['handNumber'],
            stage=d['stage'],
            you=YouState(
                name=you_d['name'],
                chips=you_d['chips'],
                hole_cards=[Card.from_str(c) for c in you_d['holeCards']],
                bet=you_d['bet'],
                position=you_d['position'],
                best_hand=BestHand.from_dict(you_d['bestHand']) if you_d.get('bestHand') else None,
            ),
            action=ActionOptions(
                can_check=act_d['canCheck'],
                to_call=act_d['toCall'],
                min_raise=act_d['minRaise'],
                max_raise=act_d['maxRaise'],
            ),
            table=TableState(
                pot=tbl_d['pot'],
                current_bet=tbl_d['currentBet'],
                community_cards=[Card.from_str(c) for c in tbl_d['communityCards']],
                small_blind=tbl_d['smallBlind'],
                big_blind=tbl_d['bigBlind'],
                ante=tbl_d['ante'],
            ),
            players=[
                Player(
                    name=p['name'], chips=p['chips'], bet=p['bet'],
                    folded=p['folded'], all_in=p['allIn'],
                    position=p['position'], disconnected=p['disconnected'],
                )
                for p in d['players']
            ],
        )


# ─────────────────────────────────────────────────────────────
# ACTIONS
# ─────────────────────────────────────────────────────────────

class Action:
    @staticmethod
    def fold() -> dict:
        return {'type': 'fold'}

    @staticmethod
    def check() -> dict:
        return {'type': 'check'}

    @staticmethod
    def call() -> dict:
        return {'type': 'call'}

    @staticmethod
    def raise_by(amount: int, opts: Optional[ActionOptions] = None) -> dict:
        if opts:
            amount = max(opts.min_raise, min(amount, opts.max_raise))
        return {'type': 'raise', 'amount': int(amount)}

    @staticmethod
    def all_in(opts: ActionOptions) -> dict:
        return {'type': 'raise', 'amount': opts.max_raise}

    @staticmethod
    def pot_sized(pot: int, opts: ActionOptions) -> dict:
        return Action.raise_by(pot, opts)

    @staticmethod
    def half_pot(pot: int, opts: ActionOptions) -> dict:
        return Action.raise_by(pot // 2, opts)

    @staticmethod
    def check_or_call(opts: ActionOptions) -> dict:
        return Action.check() if opts.can_check else Action.call()


# ─────────────────────────────────────────────────────────────
# HELPERS
# ─────────────────────────────────────────────────────────────

def pot_odds(to_call: int, pot: int) -> float:
    """Fraction of total pot you need to invest. Call if equity exceeds this."""
    if to_call == 0:
        return 0.0
    return to_call / (pot + to_call)


LATE_POSITIONS = {'BTN', 'CO', 'BTN/SB'}
MID_POSITIONS  = {'HJ', 'MP', 'MP+1'}


def pre_flop_strength(hole_cards: list, position: str) -> float:
    """
    Pre-flop hand strength 0–1, adjusted for position.
    Late position widens the playable range.
    """
    if len(hole_cards) < 2 or hole_cards[0].hidden:
        return 0.1

    cards = sorted(hole_cards, key=lambda c: c.value, reverse=True)
    a, b = cards[0], cards[1]
    is_pair    = a.rank == b.rank
    is_suited  = a.suit == b.suit
    gap        = a.value - b.value
    is_broadway = a.value >= 10 and b.value >= 10
    has_ace    = a.value == 14
    in_late    = position in LATE_POSITIONS
    in_mid     = position in MID_POSITIONS

    if is_pair:
        if a.value >= 10:  score = 0.95  # JJ+
        elif a.value >= 7: score = 0.70  # 77-TT
        else:              score = 0.50  # 22-66
    elif is_broadway:
        if a.value == 14 and b.value == 13:
            score = 0.90 if is_suited else 0.85  # AK
        elif a.value == 14:
            score = 0.80 if is_suited else 0.72  # AQ, AJ, AT
        else:
            score = 0.68 if is_suited else 0.58  # KQ, KJ, QJ
    elif has_ace:
        score = 0.60 if is_suited else 0.45      # Ax suited/offsuit
    elif is_suited and gap <= 2:
        score = 0.55 if a.value >= 7 else 0.42  # suited connectors
    elif gap == 0:
        score = 0.42 if a.value >= 8 else 0.30  # offsuit connectors
    else:
        score = 0.15  # trash

    if in_late: score = min(score + 0.12, 1.0)
    elif in_mid: score = min(score + 0.06, 1.0)

    return score


def post_flop_strength(best_hand: 'BestHand | None', opponent_count: int) -> float:
    """Post-flop strength 0–1. Discounted for more opponents."""
    if best_hand is None:
        return 0.1
    base = {
        'HIGH_CARD': 0.08, 'ONE_PAIR': 0.30, 'TWO_PAIR': 0.52,
        'THREE_OF_A_KIND': 0.65, 'STRAIGHT': 0.76, 'FLUSH': 0.82,
        'FULL_HOUSE': 0.90, 'FOUR_OF_A_KIND': 0.97,
        'STRAIGHT_FLUSH': 0.99, 'ROYAL_FLUSH': 1.0,
    }.get(best_hand.name, 0.1)
    discount = max(0, (opponent_count - 1) * 0.05)
    return max(base - discount, 0.02)


def stack_bbs(chips: int, big_blind: int) -> float:
    return chips / big_blind if big_blind else 99.0


def spr(chips: int, pot: int) -> float:
    return chips / pot if pot else 99.0


def bet_size(strength: float, pot: int, opts: ActionOptions) -> dict:
    """Scale bet size with hand strength."""
    if strength >= 0.90:   fraction = 1.00
    elif strength >= 0.70: fraction = 0.75
    elif strength >= 0.55: fraction = 0.50
    else:                  fraction = 0.33
    return Action.raise_by(int(pot * fraction), opts)


# ─────────────────────────────────────────────────────────────
# YOUR STRATEGY — implement this function
# ─────────────────────────────────────────────────────────────

def decide(state: GameState) -> dict:
    """
    Your bot's decision logic. Called every time it's your turn.
    Replace or extend this with your own strategy.
    """
    you      = state.you
    action   = state.action
    table    = state.table
    opponents = state.opponent_count()
    my_bbs   = stack_bbs(you.chips, table.big_blind)

    # ── SHORT STACK: push/fold under 10BB ────────────────────
    if my_bbs < 10 and not action.can_check:
        if pre_flop_strength(you.hole_cards, you.position) >= 0.50:
            return Action.all_in(action)
        return Action.fold()

    # ── PRE-FLOP ─────────────────────────────────────────────
    if state.is_pre_flop():
        pf_str    = pre_flop_strength(you.hole_cards, you.position)
        to_call   = action.to_call
        raise_size = table.big_blind * 2.5

        if action.can_check:
            if pf_str >= 0.55:
                return Action.raise_by(int(raise_size), action)
            return Action.check()

        call_fraction = to_call / you.chips if you.chips else 1.0
        if pf_str >= 0.85:
            return Action.raise_by(to_call * 3, action)
        if pf_str >= 0.65 and call_fraction < 0.10:
            return Action.call()
        if pf_str >= 0.50 and call_fraction < 0.05 and you.in_position():
            return Action.call()
        return Action.fold()

    # ── POST-FLOP ─────────────────────────────────────────────
    strength    = post_flop_strength(you.best_hand, opponents)
    odds        = pot_odds(action.to_call, table.pot)
    current_spr = spr(you.chips, table.pot)

    if not action.can_check:
        if strength >= 0.82 and current_spr < 4 and action.can_raise():
            return Action.all_in(action)
        if strength >= 0.72 and action.can_raise() and action.to_call < you.chips * 0.25:
            return bet_size(strength, table.pot, action)
        if strength >= odds + 0.08:
            return Action.call()
        return Action.fold()

    # Can check
    if strength >= 0.55 and action.can_raise():
        if strength >= 0.65 or you.in_position():
            return bet_size(strength, table.pot, action)
    return Action.check()

# ─────────────────────────────────────────────────────────────
# SERVER — do not modify below this line
# ─────────────────────────────────────────────────────────────

class BotHandler(BaseHTTPRequestHandler):
    def log_message(self, fmt, *args):
        pass  # suppress default access log

    def do_GET(self):
        if self.path == '/health':
            self._json(200, {'status': 'ok', 'bot': 'Python Boilerplate'})
        else:
            self.send_response(404)
            self.end_headers()

    def do_POST(self):
        if self.path == '/action':
            length = int(self.headers.get('Content-Length', 0))
            body = self.rfile.read(length)
            try:
                payload = json.loads(body)
                state = GameState.from_dict(payload)

                print(
                    f"[Hand {state.hand_number}] {state.stage.upper()} | "
                    f"{' '.join(str(c) for c in state.you.hole_cards)} | "
                    f"Pot: {state.table.pot} | To call: {state.action.to_call}"
                    + (f" | Best: {state.you.best_hand.name}" if state.you.best_hand else "")
                )

                result = decide(state)
                print(f"  → {result['type']}" + (f" {result.get('amount')}" if 'amount' in result else ""))
                self._json(200, result)

            except Exception as e:
                print(f"[Bot error] {e}")
                self._json(200, {'type': 'fold'})
        else:
            self.send_response(404)
            self.end_headers()

    def _json(self, code: int, data: dict):
        body = json.dumps(data).encode()
        self.send_response(code)
        self.send_header('Content-Type', 'application/json')
        self.send_header('Content-Length', str(len(body)))
        self.send_header('Access-Control-Allow-Origin', '*')
        self.end_headers()
        self.wfile.write(body)


if __name__ == '__main__':
    server = HTTPServer(('', PORT), BotHandler)
    print(f'🤖 Bot running on http://localhost:{PORT}')
    print(f'   POST /action  — receives game state, returns action')
    print(f'   GET  /health  — health check')
    try:
        server.serve_forever()
    except KeyboardInterrupt:
        print('\nShutting down.')
